chain/pool/src/lib.rs

```diff
@@ -1,83 +1,121 @@
+            .or_insert_with(Vec::new)
+            .push(signed_transaction);
     }
 
     /// Take transactions from the pool, in the appropriate order to be put in a new block.
-    /// Ensure that on average they will fit into expected weight.
+    /// We first take one transaction per key of (AccountId, PublicKey) with the lowest nonce,
+    /// then we take the next transaction per key with the lowest nonce.
     pub fn prepare_transactions(
         &mut self,
         expected_weight: u32,
     ) -> Result<Vec<SignedTransaction>, Error> {
-        // TODO: pack transactions better.
-        let result = self
-            .transactions
-            .values()
-            .flat_map(BTreeMap::values)
-            .take(expected_weight as usize)
-            .cloned()
-            .collect();
+        let mut sorted = false;
+        let mut result = vec![];
+        while result.len() < expected_weight as usize && !self.transactions.is_empty() {
+            let mut keys_to_remove = vec![];
+            for (key, txs) in self.transactions.iter_mut() {
+                if !sorted {
+                    // Reverse sort by nonce (non-increasing)
+                    txs.sort_by(|a, b| b.transaction.nonce.cmp(&a.transaction.nonce));
+                }
+                let tx = txs.pop().expect("transaction groups shouldn't be empty");
+                if txs.is_empty() {
+                    keys_to_remove.push(key.clone());
+                }
+                result.push(tx);
+                if result.len() >= expected_weight as usize {
+                    break;
+                }
+            }
+            sorted = true;
+            // Removing empty keys
+            keys_to_remove.into_iter().for_each(|key| {
+                self.transactions.remove(&key);
+            });
+        }
+        result.iter().for_each(|tx| {
+            self.unique_transactions.remove(&tx.get_hash());
+        });
         Ok(result)
     }
 
     /// Quick reconciliation step - evict all transactions that already in the block
     /// or became invalid after it.
-    pub fn remove_transactions(&mut self, transactions: &Vec<SignedTransaction>) {
-        for transaction in transactions.iter() {
-            let signer_id = &transaction.transaction.signer_id;
-            let nonce = transaction.transaction.nonce;
-            let mut remove_map = false;
-            if let Some(map) = self.transactions.get_mut(signer_id) {
-                map.remove(&nonce);
-                remove_map = map.is_empty();
-            }
-            if remove_map {
-                self.num_transactions -= 1;
-                self.transactions.remove(signer_id);
-            }
-        }
+    pub fn remove_transactions(&mut self, transactions: &[SignedTransaction]) {
+        transactions
+            .iter()
+            .filter(|tx| self.unique_transactions.contains(&tx.get_hash()))
+            .fold(HashMap::<_, HashSet<_>>::new(), |mut acc, tx| {
```

Comment:
This is hard to read. What are the inferred types in `HashMap::<_, HashSet<_>>`? What is the meaning behind this temporary accumulator? What is the type of `key` in `for_each(|(key, hashes)| {`? I don't think this piece of code benefits from being functional. The logic seems to be very procedural.